src/typecheck.rs

```diff
@@ -734,36 +726,106 @@ fn type_check_(
+            | ApparentType::Approximated(ty) => to_typewrapper(ty),
+            ApparentType::FromEnv(tyw) => tyw,
+        }
     }
 }
 
 /// Determine the apparent type of a let-bound expression.
 ///
 /// When a let-binding `let x = bound_exp in body` is processed, the type of `bound_exp` must be
-/// determined to be associated to the bound variable `x` in the typing environment (`typed_vars`).
+/// determined in order to be bound to the variable `x` in the typing environment.
 /// Then, future occurrences of `x` can be given this type when used in a `Promise` block.
 ///
 /// The role of `apparent_type` is precisely to determine the type of `bound_exp`:
-/// - if `bound_exp` is annotated by an `Assume` or a `Promise`, return the user-provided type.
-/// - Otherwise, `None` is returned.
-pub fn apparent_type(t: &Term) -> Option<Types> {
+/// - if `bound_exp` is annotated by an `Assume`, a `Promise` or a metavalue, return the
+///   user-provided type.
+/// - if `bound_exp` is a constant (string, number, boolean or symbol) which type can be deduced
+///   directly without unfolding the expression further, return the corresponding exact type.
+/// - if `bound_exp` is a list, return `List Dyn`.
+/// - Otherwise, return an approximation of the type (currently `Dyn`, but could be more precise in
+///   the future, such as `Dyn -> Dyn` for functions, `{ | Dyn}` for records, and so on).
+pub fn apparent_type(t: &Term, envs: Option<&Envs>) -> ApparentType {
+    fn is_literal(chunk: &StrChunk<RichTerm>) -> bool {
+        match chunk {
+            StrChunk::Literal(_) => true,
+            _ => false,
+        }
+    }
+
     match t {
-        Term::Assume(ty, _, _) | Term::Promise(ty, _, _) => Some(ty.clone()),
+        Term::Assume(ty, _, _) | Term::Promise(ty, _, _) => ApparentType::Annotated(ty.clone()),
         Term::MetaValue(MetaValue {
             contract: Some((ty, _)),
             ..
-        }) => Some(ty.clone()),
+        }) => ApparentType::Annotated(ty.clone()),
         Term::MetaValue(MetaValue {
             contract: None,
             value: Some(v),
             ..
-        }) => apparent_type(v.as_ref()),
-        _ => None,
+        }) => apparent_type(v.as_ref(), envs),
+        Term::Num(_) => ApparentType::Exact(Types(AbsType::Num())),
+        Term::Bool(_) => ApparentType::Exact(Types(AbsType::Bool())),
+        Term::Sym(_) => ApparentType::Exact(Types(AbsType::Sym())),
+        Term::Str(_) => ApparentType::Exact(Types(AbsType::Str())),
+        // If there are interpolated expressions inside a string, we can't guarantee that this is a
+        // well-typed string (e.g. "#{1 + 1}" is not). We can only do so if all chunks are string
+        // literals.
+        Term::StrChunks(chunks) if chunks.iter().all(is_literal) => {
+            ApparentType::Exact(Types(AbsType::Str()))
+        }
+        Term::List(_) => ApparentType::Exact(Types(AbsType::List(Box::new(Types(AbsType::Dyn()))))),
```

Comment:
It's a little bit weird to call this case `Exact` isn't it?

Reply:
Yeah, it was in opposition to `Approximated`. Maybe `Precise`, or `Inferred`?